### ISSUE TITLE
[feature request] Support direct apply for icon packs

### ISSUE THREAD

**Issue by Donnnno** [MM2-0/Kvaesitso#395]
Some big Launchers have options to apply an icon pack from the pack itself (see screenshots), this would be nice for icon pack users to have. 

There needs to be an implementation from the Launcher & the icon pack for this tho, here's how Candybar (the icon dashboard) handles it:
https://github.com/zixpo/candybar/blob/529a7619c4e03e814ddf4430258ff778cf734159/library/src/main/java/candybar/lib/helpers/LauncherHelper.java#LL460C1-L476C23

https://github.com/zixpo/candybar/blob/529a7619c4e03e814ddf4430258ff778cf734159/library/src/main/java/candybar/lib/helpers/LauncherHelper.java#L477C1-L497

![image](https://github.com/MM2-0/Kvaesitso/assets/31142286/2ac7b264-4053-4257-b2a5-43169b4fdcb2) ![image](https://github.com/MM2-0/Kvaesitso/assets/31142286/c7065ecf-bc97-4560-adfc-151835fba25a)

**Comment by MM2-0** [comment]
> https://github.com/zixpo/candybar/blob/529a7619c4e03e814ddf4430258ff778cf734159/library/src/main/java/candybar/lib/helpers/LauncherHelper.java#L477C1-L497

If I read this correctly, every launcher has its own API to do this and every icon pack needs explicit support for every single launcher. I kinda dislike that. I would be down to add this if there was an open standard for this though. I guess i could hijack Niagara's or Lawnchair's intent as they don't seem to use a hardcoded component or package name but that would be pretty unintuitive.

**Comment by Donnnno** [comment]
Totally understandable! 
It's annoying that there isn't a standard for this, so the solution is kinda wonky.

For now, I've implemented it to look like this with Kvaesitso: 

![afbeelding](https://github.com/MM2-0/Kvaesitso/assets/31142286/ca2f8e1a-413a-43dd-b1be-2798959a1ebc)

**Comment by MM2-0** [comment]
> It's annoying that there isn't a standard for this, so the solution is kinda wonky.

I guess we could create such a standard but we'd need to make sure that Candybar adds it using PackageManager.queryIntentActivity instead of hardcoding yet another launcher package name.

**Comment by Donnnno** [comment]
> I guess we could create such a standard but we'd need to make sure that Candybar adds it using PackageManager.queryIntentActivity instead of hardcoding yet another launcher package name.

Sounds cool, though I have no idea how any of this works 😅 
I could ask the main Candybar dev what he thinks of that, and if something like that would be possible.

**Comment by moertel** [comment]
CandyBar contributor and icon pack developer here. Thanks for starting the discussion, @Donnnno!

FWIW, when Nova introduced support for dynamic calendar icons[1], they chose to go with their own namespace and declare it an open standard that other launchers should be adopting too. Which worked, surprisingly.

> I guess we could create such a standard but we'd need to make sure that Candybar adds it using PackageManager.queryIntentActivity instead of hardcoding yet another launcher package name.

I'd be happy to contribute this to CandyBar and make sure it's implemented, tested and merged.

[1] https://github.com/teslacoil/Example_NovaTheme/blob/master/DynamicCalendarIconAPI.md

**Comment by MM2-0** [comment]
> FWIW, when Nova introduced support for dynamic calendar icons[1], they chose to go with their own namespace and declare it an open standard that other launchers should be adopting too. Which worked, surprisingly.

True, but Nova has a greater market power so it's easier for them to establish new standards.

> I'd be happy to contribute this to CandyBar and make sure it's implemented, tested and merged.

That would be great!

> Adding support to these two core libraries will automatically enable support for all icon packs that use them.

As long as they update the library I suppose. But that simplifies the situation by a lot.

Talking about establishing new standards: I'd also like to establish a new standard for dynamic clock icons in icon packs (similar to the Nova calendar icon standard).

**Comment by Donnnno** [comment]
For dynamic clocks there's this: https://medium.com/@sarsamurmu/adaptive-icons-in-icon-pack-1efdc1f77545#6efd

But never used it myself because it doesn't work with transparent icons :(

**Comment by MM2-0** [comment]
> For dynamic clocks there's this: https://medium.com/@sarsamurmu/adaptive-icons-in-icon-pack-1efdc1f77545#6efd
> 
> But never used it myself because it doesn't work with transparent icons :(

Thanks, I didn't know about this, but I will add support for it. Which other launcher support that?

**Comment by Donnnno** [comment]
Nova for sure

**Comment by MM2-0** [comment]
@Donnnno I have added support for dynamic clock icons in icon packs (it will be available in nightlies tomorrow). I have also written [some docs](https://kvaesitso.mm20.de/docs/developer-guide/integrations/icon-packs/#clock-icons) for it. I made it so that you can use a plain `layer-list` for the clock icon, but unfortunately this does not work with other launchers.

**Comment by Donnnno** [comment]
Oh cool! I'm gonna test this tomorrow!

**Comment by Donnnno** [comment]
Hey @MM2-0 ! I've tried it, but no luck yet... Did I miss something?

https://github.com/Donnnno/Arcticons/commit/0e5a361dc4a59acef91a3120bd28f37a2036255a

**Comment by MM2-0** [comment]
I think the issue is that density resource qualifiers have higher precedence than API version qualifiers. The `v26` icon variant is overridden by the `nodpi` variant.

**Comment by Donnnno** [comment]
It's working on the emulator! (my Samsung phone doesn't work tho)

![image](https://github.com/MM2-0/Kvaesitso/assets/31142286/537c834b-c09c-4eda-b290-c847c65c92c7)

But tested it with nova, and it has a black background :( 

![image](https://github.com/MM2-0/Kvaesitso/assets/31142286/6f139912-7ad1-4267-bf95-436cc8fd2bbe) (light mode for dramatic effect)

So gotta find a solution for that part (because if other popular launchers render it wrong, it might not be the best idea to implement it at Arcticons, unfortunately.

**Comment by MM2-0** [comment]
If you use a plain layer-list (instead of an adaptive icon), Nova renders it as a static icon. So no dynamic clock unfortunately, but at least not broken.

**Comment by Donnnno** [comment]
> If you use a plain layer-list (instead of an adaptive icon), Nova renders it as a static icon. So no dynamic clock unfortunately, but at least not broken.
Sounds logical, but for some reason, doing that makes the normal icon super tiny 

![image](https://github.com/MM2-0/Kvaesitso/assets/31142286/4792e531-0519-4b4c-8ae2-2cd1c6d49d48)

**Comment by MM2-0** [comment]
Your scaling is wrong.
[This icon](https://raw.githubusercontent.com/Donnnno/Arcticons/0e5a361dc4a59acef91a3120bd28f37a2036255a/app/src/dark/res/drawable-nodpi/clock_top.png) has way too much padding compared to [this](https://raw.githubusercontent.com/Donnnno/Arcticons/0e5a361dc4a59acef91a3120bd28f37a2036255a/app/src/dark/res/drawable-nodpi/clock.png)

**Comment by Donnnno** [comment]
Oops, that was the adaptive icon scaling, thank you so much! It works perfect now!

**Comment by moertel** [comment]
I'm glad you solved the clock thing. Will the direct apply for icons still be considered / implemented?
